internal/indexer/postgres/affectedmanifest.go

```diff
@@ -0,0 +1,269 @@
+package postgres
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/rs/zerolog"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/pkg/omnimatcher"
+)
+
+var (
+	// ErrNotIndexed indicates the vulnerability being queried has a dist or repo not
+	// indexed into the database.
+	ErrNotIndexed = fmt.Errorf("vulnerability containers data not indexed by any scannners")
+)
+
+// affectedManifests finds the manifests digests which are affected by the provided vulnerability.
+//
+// an exhaustive search for all indexed packages of the same name as the vulnerability is performed.
+//
+// the list of packages is filtered down to only the affected set.
+//
+// the manifest index is then queried to resolve a list of manifest hashes containing the affected
+// artifacts.
+func affectedManifests(ctx context.Context, pool *pgxpool.Pool, v claircore.Vulnerability) ([]claircore.Digest, error) {
+	const (
+		selectPackages = `
+		SELECT id, name, version, kind
+		FROM package
+		WHERE name = $1;
+		`
+		selectAffected = `
+		SELECT manifest_hash
+		FROM manifest_index
+		WHERE package_id = $1
+		  AND (
+			CASE
+				WHEN $2::bigint IS NULL THEN
+					dist_id IS NULL
+				ELSE
+					dist_id = $2
+				END
+			)
+		  AND (
+			CASE
+				WHEN $3::bigint IS NULL THEN
+					repo_id IS NULL
+				ELSE
+					repo_id = $3
+				END
+			);
```

Comment:
Clever

Reply:
gotta thanks the IRC folks for this one.